src/main/java/com/google/sps/servlets/MealServlet.java

```diff
@@ -93,13 +96,34 @@ private Entity createMealEntity(Meal meal) {
     }
 
     private void getMealList(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        String pathInfo = request.getPathInfo();
+        if (pathInfo == null) {
+            pathInfo = "";
+        }
+        List<String> params = new ArrayList<>(
+            Arrays.asList(pathInfo.replaceAll("/query=", "").trim().replaceAll("\\s+", " ").split(" ")));
         Query query = new Query("Meal").addSort("id", SortDirection.ASCENDING);
         DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
         PreparedQuery results = datastore.prepare(query);
         List<Meal> meals = getDataFromDatastore(results);
+        List<Meal> searchedMeal = new ArrayList<>();
+
+        // Every object Meal adds if it contains at least one keyword in at least one field 
+        // (type, title, description, ingredients).
+        for (Meal meal : meals) {
+            for (String param : params) {
+                if (!searchedMeal.contains(meal) && 
```

Comment:
Probably it will be better to create func `isRelated` for example, in which you will have all this conditions.

Reply:
It's a good idea to put this big condition into a function that will act as a predicate. It should run against arguments `Meal` and `Collection<String>`, so that the check against multiple query params is contained within the function (instead of using `.contains` to check if the meal already classified as matching the query. `contains` for an `List` is of linear complexity by the way, so we should avoid it to not increase the complexity of the search algorithm).

In a perfect solution, we'd use Elasticsearch to actually perform the search, instead of our own application filtering the results out. However, this requires learning new API, some tricky setup (besides starting an Elasticsearch engine some workarounds for the API is needed on App Engine) and a lot of extra effort that we might not afford (especially if we want some other features to be implemented before internship ends).

So let's keep it this way, but keep in mind - in a perfect world, such filtering would be performed by the database (or some dedicated search engine) to make the search efficient (i.e. not browse the entire database every time, but use indexes to optimize the search by a *lot*).